### PR TITLE
Daisy Initial Kit Code

### PR DESCRIPTION
**Pull request by iamtesch**
This code should still be cleaned up quite a bit to improve quality, but this initial PR acts as a functional demonstration of hexapod code in C++.

### CODE REVIEW

File: kits/daisy/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+# Get the C++ API directory:
+set(ROOT_DIR ${CMAKE_CURRENT_SOURCE_DIR}/../..)
+get_filename_component(ROOT_DIR "${ROOT_DIR}" REALPATH)
+# `HEBI_CPP_DIR` can be used to point to an already downloaded C++ API.
+# This is useful for environments that would prefer not to download files
+# from the internet (or even don't have internet access).
+if(NOT HEBI_CPP_DIR)
+  set(HEBI_DIR ${ROOT_DIR}/hebi-cpp)
+  include(../../projects/cmake/DownloadHebiCpp.cmake)
```
Comment (rLinks234) [inline kits/daisy/CMakeLists.txt]: Why not use `ROOT_DIR` which was created above instead of assuming the working directory is `${CMAKE_CURRENT_SOURCE_DIRECTORY}`?

File: kits/daisy/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+get_filename_component(hebi_cpp_build_dir "${CMAKE_CURRENT_BINARY_DIR}/${hebi_cpp_build_dir}" REALPATH)
+add_subdirectory(${HEBI_DIR} ${hebi_cpp_build_dir})
+
+include_directories (
+  ${CMAKE_CURRENT_SOURCE_DIR}/src          # Our source files
+  ${HEBI_DIR}/src ${HEBI_DIR}/hebi/include ${HEBI_DIR}/Eigen)
+link_directories (
+  ${HEBI_CPP_LINK_DIRECTORIES})
+
+add_library(hexapod_core OBJECT
```
Comment (rLinks234) [inline kits/daisy/CMakeLists.txt]: We need to stop doing these `OBJECT` libraries and adhere to Windows' strange standard of requiring `__declspec(dllexport)` on exported elements

File: kits/daisy/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+#qt5_use_modules(parameter_gui Core Gui Widgets 3DCore 3DExtras 3DRender 3DInput)
+#target_link_libraries(parameter_gui ${SONAME} hebi hebic++ m pthread)
+
+# 2D Parameter setting GUI
+add_executable(parameter_gui_2d ${PARAMETER_GUI_2D_SOURCES} $<TARGET_OBJECTS:hexapod_core>)
+qt5_use_modules(parameter_gui_2d Core Gui Widgets)
+target_link_libraries(parameter_gui_2d ${SONAME} hebi hebic++ m pthread)
+
+add_executable(input_test ${INPUT_TEST_SOURCES})
+target_link_libraries( input_test ${SONAME} hebi hebic++ m pthread )
```
Comment (rLinks234) [inline kits/daisy/CMakeLists.txt]: I unfortunately remember that this won't work on Windows, because Windows `hebic++` doesn't actually export any functions (no `__declspec(dllexport)` on anything due to obvious oversight on my part)

File: kits/daisy/src/input/input_manager_mobile_io.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  hebi::Lookup lookup;
+  long timeout_ms = 2000; // use a 2 second timeout
+  group_ = lookup.getGroupFromNames({"HEBI"}, {"Mobile IO"}, timeout_ms);
+  if (!group_)
+    return false;
+
+  group_->addFeedbackHandler([this] (const GroupFeedback& fbk)
+  {
+    // TODO: use 'std::atomic' variables for state here? Add mutex?  Does this
+    // slow things down due to lock contention?
```
Comment (rLinks234) [inline kits/daisy/src/input/input_manager_mobile_io.cpp]: There are a lot of other things which need to be cleaned up which will result in better performance than lock contention here

File: kits/daisy/src/input/input_manager_mobile_io.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        ++num_mode_toggles_;
+      prev_mode_button_state_ = new_mode_button_state;
+
+      // Check for "Quit"
+      if (digital.getInt(8) == 1) {
+        has_quit_been_pushed_ = true;
+      }
+    }
+  });
+  group_->setFeedbackFrequencyHz(1000); // Don't want to miss button presses!
```
Comment (rLinks234) [inline kits/daisy/src/input/input_manager_mobile_io.cpp]: I don't think turning the frequency up all of the way will result in a better chance of getting button presses, and risks just dropping a ton more packets because wifi isn't going to handle this well whatsoever. This is also problematic at the user facing side, since it portrays the assumption that it's going to work well like this

File: kits/daisy/src/input/input_manager_mobile_io.hpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+  // The Mobile IO app that serves as a joystick
+  std::shared_ptr<hebi::Group> group_;
+
+  // Scale the joystick scale to motion of the robot in SI units (m/s, rad/s,
+  // etc).
+  static constexpr float xyz_scale_ = 0.175;
+  static constexpr float rot_scale_ = 0.4;
+
+  float left_horz_raw_ = 0; // Rotation
```
Comment (rLinks234) [inline kits/daisy/src/input/input_manager_mobile_io.hpp]: brace initialize all of these

File: kits/daisy/src/main.cpp [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+  partial = false;
+  quiet = false;
+  bool valid = true;
+  int idx = 1; // Ignore program name!
+  for (;idx < argc; ++idx)
+  {
+    std::string str_arg(argv[idx]);
+    if (str_arg == "-h")
+    {
+      std::cout << "Hexapod control usage:" << std::endl <<
```
Comment (rLinks234) [inline kits/daisy/src/main.cpp]: `std::endl`

File: kits/daisy/src/main.cpp [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+      for (++idx; idx < argc; ++idx)
+      {
+        try
+        {
+          int partial_leg = std::stoi(std::string(argv[idx]));
+          partial_legs.insert(partial_leg);
+        }
+        catch (const std::invalid_argument&)
+        {
+          idx -= 2; // -2, because we will add one next start of the parent loop
```
Comment (rLinks234) [inline kits/daisy/src/main.cpp]: How much have you tested this logic?

File: kits/daisy/src/main.cpp [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+  }
+
+  std::unique_ptr<Hexapod> hexapod = getHexapod(params, is_dummy, is_partial, is_quiet, legs);
+  while (is_quiet && !hexapod)
+  {
+    hexapod = getHexapod(params, is_dummy, is_partial, is_quiet, legs);
+  }
+
+  if(!hexapod && !is_quiet)
+    return -1;
```
Comment (rLinks234) [inline kits/daisy/src/main.cpp]: Return codes are practically never negative, so change to `+1`.

See [this](https://unix.stackexchange.com/questions/418784/what-is-the-min-and-max-values-of-exit-codes-in-linux) for more

File: kits/daisy/src/main.cpp [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+    vLayout->addLayout(hLayout2);
+    vLayout->addLayout(hLayout3);
+  }
+
+  widget->setWindowTitle(QStringLiteral("heXapod Control"));
+  widget->show();
+  widget->resize(overall_width, overall_height);
+
+  Eigen::Vector3f translation_velocity_cmd;
+  translation_velocity_cmd << 0, 0, 0;
```
Comment (rLinks234) [inline kits/daisy/src/main.cpp]: I really despise this syntax with Eigen. Why not just `setZero()`? This was a big pain point for me when I first starting using it, and the `,` operator is just pure evil

File: kits/daisy/src/main.cpp [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+        // trajectory, with zero vel/accel endpoints.
+        for (int i = 0; i < 6; ++i)
+        {
+          Eigen::VectorXd leg_start = hexapod->getLegFeedback(i);
+          Eigen::VectorXd leg_end;
+          Eigen::VectorXd leg_vels;
+          Eigen::MatrixXd jacobian_ee(0, 0);
+          robot_model::MatrixXdVector jacobian_com;
+          hexapod->getLeg(i)->computeState(elapsed.count(), leg_end, leg_vels, jacobian_ee, jacobian_com);
+          // TODO: fix! (quick and dirty -- leg mid is hardcoded as offset from leg end)
```
Comment (rLinks234) [inline kits/daisy/src/main.cpp]: Don't forget to fix

File: kits/daisy/src/main.cpp [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+        }
+
+        // TODO: move this startup logic outside of t.elapse() call, and then
+        // we don't need 'first_run'?
+        first_run = false;
+      }
+      mode->setText("Startup");
+     
+      // Follow t_l:
+      // TODO
```
Comment (rLinks234) [inline kits/daisy/src/main.cpp]: TODO what?

File: kits/daisy/src/robot/hexapod.cpp [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+  // Get leg offset, taking into account all legs may not be present. First,
+  // count the number of legs before this one:
+  int legs_prev = std::count_if(real_legs_.begin(), real_legs_.end(),
+    [leg_index](int real_leg_idx) {return real_leg_idx < leg_index; });
+  int leg_offset = legs_prev * num_joints;
+  if (angles != nullptr)
+  {
+    assert(angles->size() == num_joints);
+    for (int i = 0; i < num_joints; ++i)
+      cmd_[leg_offset + i].actuator().position().set((*angles)[i]);
```
Comment (rLinks234) [inline kits/daisy/src/robot/hexapod.cpp]: Don't we have a way to just do `cmd_[leg_offset+i].setPosition(*angles)` now?

File: kits/daisy/src/robot/hexapod.cpp [hunk 14]
```diff
@@ -0,0 +1,610 @@
...
+  }
+  if (vels != nullptr)
+  {
+    assert(vels->size() == num_joints);
+    for (int i = 0; i < num_joints; ++i)
+      cmd_[leg_offset + i].actuator().velocity().set((*vels)[i]);
```
Comment (rLinks234) [inline kits/daisy/src/robot/hexapod.cpp]: See comment above, also see above for `effort()` below

File: kits/daisy/src/robot/hexapod.cpp [hunk 15]
```diff
@@ -0,0 +1,610 @@
...
+    foot_forces.block<3,1>(0,i) = factors(i) * weight_ * grav;
+}
+
+// TODO: make this a class function and remove parameters? Make it better!
+bool copyIntoPositions(Eigen::VectorXd& positions, const hebi::GroupFeedback* fbk, const std::set<int>& real_legs)
+{
+  bool valid_fbk = true;
+  int fbk_idx = 0;
+  int num_leg_modules = 3;
+  for (int i = 0; i < positions.size();)
```
Comment (rLinks234) [inline kits/daisy/src/robot/hexapod.cpp]: I really don't understand what's going on here

File: kits/daisy/src/robot/hexapod.cpp [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+  last_step_legs_.insert(4);
+
+  last_fbk = std::chrono::steady_clock::now();
+  // Start a background feedback handler
+  if (group_)
+  {
+    group->addFeedbackHandler([this] (const GroupFeedback& fbk)
+    {
+      Eigen::Vector3d avg_accel;
+      avg_accel << 0, 0, 0; // unnecessary?
```
Comment (rLinks234) [inline kits/daisy/src/robot/hexapod.cpp]: `setZero`

File: kits/daisy/src/robot/hexapod.cpp [hunk 17]
```diff
@@ -0,0 +1,610 @@
...
+        int num_prev_legs = std::count_if(real_legs_.begin(), real_legs_.end(), [i](int other_leg) { return other_leg < i; });
+        auto accel = fbk[num_prev_legs * num_leg_joints].imu().accelerometer().get(); 
+        auto gyro = fbk[num_prev_legs * num_leg_joints].imu().gyro().get(); 
+        Eigen::Vector3d mod_accel;
+        mod_accel << accel.getX(), accel.getY(), accel.getZ();
+        Eigen::Vector3d mod_gyro;
+        mod_gyro << gyro.getX(), gyro.getY(), gyro.getZ();
+        // Transform
+        Eigen::Matrix4d trans = legs_[i]->getKinematics().getBaseFrame();
+        avg_accel += trans.topLeftCorner(3,3) * mod_accel;
```
Comment (rLinks234) [inline kits/daisy/src/robot/hexapod.cpp]: `trans.topLeftCorner<3, 3>()` gives better compile time opportunities for codegen

File: kits/daisy/src/robot/hexapod.cpp [hunk 18]
```diff
@@ -0,0 +1,610 @@
...
+    log_group_modules_->startLog("logs", log_name_base + "-HEX.hebilog");
+
+  }
+}
+
+Hexapod::~Hexapod()
+{
+  if (group_)
+  {
+    group_->setFeedbackFrequencyHz(0);
```
Comment (rLinks234) [inline kits/daisy/src/robot/hexapod.cpp]: This is already done in their constructors?

File: kits/daisy/src/robot/hexapod.hpp [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+//     / \      +y <--o
+//  5 /   \ 6          +z
+//
+// Distances based on CAD on October 17 2018
+
+using namespace Eigen;
+
+namespace hebi {
+
+typedef struct {
```
Comment (rLinks234) [inline kits/daisy/src/robot/hexapod.hpp]: This is C++...

File: kits/daisy/src/robot/leg.cpp [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+    std::cerr << "Could not find or load HRDF file for leg!" << std::endl;
+    assert("false");
+    return;
+  }
+
+  kin_->getMasses(masses_);
+
+  Matrix4d transform = Matrix4d::Identity();
+  Matrix3d rotate = AngleAxisd(angle_rad, Eigen::Vector3d::UnitZ()).matrix();
+  transform.topLeftCorner<3,3>() = rotate;
```
Comment (rLinks234) [inline kits/daisy/src/robot/leg.cpp]: :+1: for template overload of this (`<3, 3>`)

File: kits/daisy/src/robot/step.cpp [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+    leg_waypoint_vels.col(next_pt).setZero();
+    leg_waypoint_accels.col(next_pt).setConstant(std::numeric_limits<double>::quiet_NaN());
+    leg_times.push_back(time_[3]);
+    next_pt++;
+  }
+*/
+  int num_pts = next_pt;
+  // TODO: don't resize, but just be smarter about passing in partial vectors
+  // or creating the right size in the first place!
+  leg_waypoint_vels.conservativeResize(num_joints, num_pts);
```
Comment (rLinks234) [inline kits/daisy/src/robot/step.cpp]: Or better yet, revamp our API to allow for known matrix sizes at compile time (e.g. `Eigen::Matrix<double, 6, 3>`). This is a pain point I saw countless times while porting Igor to C++

File: kits/daisy/src/robot/step.hpp [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+  // Step waypoints: lift up -> high point -> touch down
+  Eigen::Vector3d lift_up_;
+  Eigen::Vector3d mid_step_1_;
+  Eigen::Vector3d mid_step_2_;
+  Eigen::Vector3d touch_down_;
+
+  std::shared_ptr<trajectory::Trajectory> trajectory_;
+  
+//  static constexpr float
```
Comment (rLinks234) [inline kits/daisy/src/robot/step.hpp]: ?

File: kits/daisy/src/xml_util/xml_helpers.hpp [hunk 23]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include "pugixml.hpp"
+
+namespace hebi {
+
+class XMLHelpers
```
Comment (rLinks234) [inline kits/daisy/src/xml_util/xml_helpers.hpp]: Why is this here? Why aren't they free standing functions?

Comment (rLinks234) [reply kits/daisy/src/xml_util/xml_helpers.hpp]: If you really need them to be scoped, there's no reason to just place them in a namespace like `hebi::xml`. Putting them in a class where there are only `static` functions is confusing

File: kits/daisy/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,136 @@
+cmake_minimum_required (VERSION 3.1)
+project (hexapod)
+
+SET (CMAKE_CXX_STANDARD 11)
+SET (CMAKE_CXX_STANDARD_REQUIRED ON)
+SET (CMAKE_CXX_FLAGS "-O3 -g")
```
Comment (rLinks234) [inline kits/daisy/CMakeLists.txt]: I'd prefer for all of these flags to be set per target

File: kits/daisy/CMakeLists.txt (hunk 1 above)
Comment (iamtesch) [reply kits/daisy/CMakeLists.txt]: In regards to all the CMake comments -- completely agree this needs a pass.  I took the existing CMake file, and updated it to use the common `hebi-cpp` folder, but I didn't do any other cleanup yet.

FWIW, this code hasn't been made/tested as Windows compliant at all yet; I'll add an issue for that + the cmake refactoring.

File: kits/daisy/src/input/input_manager_mobile_io.cpp (hunk 5 above)
Comment (iamtesch) [reply kits/daisy/src/input/input_manager_mobile_io.cpp]: Previously, this hadn't been going over wifi -- I hadn't thought about this since that change was made.

In this case, wouldn't wifi just buffer the packets instead of dropping them?  I feel like this would still make it more likely to catch a momentary button press that happened for a fraction of a second, although I agree that this isn't as ideal as when wired.

Overall, for this it would be nice to just use a "toggle button" instead of a "momentary" switch on the I/O app -- I'll add an issue to address this.

File: kits/daisy/src/main.cpp (hunk 7 above)
Comment (iamtesch) [reply kits/daisy/src/main.cpp]: heh...yeah, sorry.  I'll fix.

File: kits/daisy/src/main.cpp (hunk 8 above)
Comment (iamtesch) [reply kits/daisy/src/main.cpp]: quite a bit

Comment (iamtesch) [reply kits/daisy/src/main.cpp]: Still not pretty; it might be worth using a header-only argument parsing library in place of this -- we had one in the old C API examples that I liked.

File: kits/daisy/src/main.cpp (hunk 10 above)
Comment (iamtesch) [reply kits/daisy/src/main.cpp]: Fixed.

File: kits/daisy/src/robot/hexapod.cpp (hunk 13 above)
Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.cpp]: Nope -- because this is just setting data for a subset of `cmd_`; we are setting one leg at a time here (hence computing all the offset).

In MATLAB, this is really easy to do with a sort of "mask" on the elements you want to set, and you can just set partial vectors (e.g., elements 1-3 of the 18 element position array).  We could add a helper utility for that in the C++ API or even just in the hexapod code, but right now this is the only way :(

File: kits/daisy/src/robot/hexapod.cpp (hunk 14 above)
Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.cpp]: See explanation above.

File: kits/daisy/src/robot/hexapod.cpp (hunk 15 above)
Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.cpp]: Added a comment to clarify.  Let's discuss -- we could potentially remove all of the "partial number of legs" support to clarify a decent bit of code.

File: kits/daisy/src/robot/hexapod.cpp (hunk 16 above)
Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.cpp]: fixed.

File: kits/daisy/src/robot/hexapod.cpp (hunk 17 above)
Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.cpp]: fixed.

File: kits/daisy/src/robot/hexapod.cpp (hunk 18 above)
Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.cpp]: It didn't used to be :(

Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.cpp]: Also, this is the destructor.

File: kits/daisy/src/input/input_manager_mobile_io.cpp (hunk 5 above)
Comment (rLinks234) [reply kits/daisy/src/input/input_manager_mobile_io.cpp]: UDP would potentially buffer the packets, but that is entirely implementation specific. Phones probably don't even acknowledge touch events at 1000Hz (it would only make sense for them to ack them at the 60Hz the screen is locked rendering at)

File: kits/daisy/src/main.cpp (hunk 8 above)
Comment (rLinks234) [reply kits/daisy/src/main.cpp]: Good enough for me

File: kits/daisy/src/main.cpp (hunk 12 above)
Comment (rLinks234) [reply kits/daisy/src/main.cpp]: Ship it!!

File: kits/daisy/src/robot/hexapod.cpp (hunk 18 above)
Comment (rLinks234) [reply kits/daisy/src/robot/hexapod.cpp]: I meant destructor, sorry

File: kits/daisy/CMakeLists.txt (hunk 1 above)
Comment (rLinks234) [reply kits/daisy/CMakeLists.txt]: re: Windows, that's completely fine. We can fix that in the upcoming weeks.

File: kits/daisy/CMakeLists.txt (hunk 2 above)
Comment (rLinks234) [reply kits/daisy/CMakeLists.txt]: This was mainly as a comment; there's nothing to resolve with this right now

File: kits/daisy/src/robot/hexapod.hpp (hunk 19 above)
Comment (iamtesch) [reply kits/daisy/src/robot/hexapod.hpp]: fixed

File: kits/daisy/src/robot/leg.cpp (hunk 20 above)
Comment (iamtesch) [reply kits/daisy/src/robot/leg.cpp]: I'm not always sloppy ;)

File: kits/daisy/CMakeLists.txt (hunk 3 above)
Comment (rLinks234) [reply kits/daisy/CMakeLists.txt]: This is basically a comment, so there's nothing to resolve here ATM

File: kits/daisy/src/input/input_manager_mobile_io.cpp (hunk 4 above)
Comment (rLinks234) [reply kits/daisy/src/input/input_manager_mobile_io.cpp]: Again, a comment that doesn't really need any resolution ATM

File: kits/daisy/src/robot/step.cpp (hunk 21 above)
Comment (iamtesch) [reply kits/daisy/src/robot/step.cpp]: I'll flag this as an issue on the C++ API

File: kits/daisy/src/robot/step.hpp (hunk 22 above)
Comment (iamtesch) [reply kits/daisy/src/robot/step.hpp]: I can't believe you don't like my random meaningless comments.

File: kits/daisy/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+SET (SONAME "hebi")
```
[the remark this answers is not in the document]
Comment (iamtesch) [reply kits/daisy/CMakeLists.txt]: fixed.

File: kits/daisy/src/xml_util/xml_helpers.hpp (hunk 23 above)
Comment (iamtesch) [reply kits/daisy/src/xml_util/xml_helpers.hpp]: I think this came from an ugly period in my life where I coded in Java.  I profusely apologize.

File: kits/daisy/src/input/input_manager_mobile_io.cpp (hunk 5 above)
Comment (iamtesch) [reply kits/daisy/src/input/input_manager_mobile_io.cpp]: Will fix ASAP
